google/cloud/pubsub/message.h

```diff
@@ -0,0 +1,212 @@
+  std::string const& ordering_key() const { return proto_.ordering_key(); }
+  std::chrono::system_clock::time_point publish_time() const;
+  std::map<std::string, std::string> attributes() const {
+    std::map<std::string, std::string> r;
+    for (auto const& kv : proto_.attributes()) {
+      r.emplace(kv.first, kv.second);
+    }
+    return r;
+  }
+  //@}
+
+  /// @name Copy and move
+  //@{
+  Message(Message const&) = default;
+  Message& operator=(Message const&) = default;
+  Message(Message&&) = default;
+  Message& operator=(Message&&) = default;
+  //@}
+
+  /// @name Equality operators
+  //@{
+  friend bool operator==(Message const& a, Message const& b);
+  friend bool operator!=(Message const& a, Message const& b) {
+    return !(a == b);
+  }
+  //@}
+
+  /// Output in protobuf format, this is intended for debugging
+  friend std::ostream& operator<<(std::ostream& os, Message const& rhs);
+
+ private:
+  friend Message pubsub_internal::FromProto(
+      ::google::pubsub::v1::PubsubMessage m);
+  friend ::google::pubsub::v1::PubsubMessage const& pubsub_internal::ToProto(
+      Message const& m);
+  friend ::google::pubsub::v1::PubsubMessage&& pubsub_internal::ToProto(
+      Message&& m);
+  friend class MessageBuilder;
+
+  Message() = default;
+
+  explicit Message(::google::pubsub::v1::PubsubMessage m)
+      : proto_(std::move(m)) {}
+
+  google::pubsub::v1::PubsubMessage proto_;
+};
+
+/**
+ * Constructs `Message` objects.
+ */
+class MessageBuilder {
+ public:
+  MessageBuilder() = default;
+
+  /// Create a new message.
+  Message Build() && { return Message(std::move(proto_)); }
+
+  /// Create a message with the data in @p data
+  MessageBuilder& SetData(std::string data) & {
+    proto_.set_data(std::move(data));
+    return *this;
+  }
+
+  /// Create a message with the data in @p data
+  MessageBuilder&& SetData(std::string data) && {
+    (*this).SetData(std::move(data));
+    return std::move(*this);
+  }
+
+  /// Create a message with the attributes from the range [@p begin, @p end)
+  template <typename Iterator>
+  MessageBuilder& SetAttributes(Iterator begin, Iterator end) & {
+    google::protobuf::Map<std::string, std::string> tmp;
+    using value_type =
+        google::protobuf::Map<std::string, std::string>::value_type;
+    for (auto kv = begin; kv != end; ++kv) {
+      using std::get;
+      tmp.insert(value_type(get<0>(*kv), get<1>(*kv)));
+    }
+    proto_.mutable_attributes()->swap(tmp);
+    return *this;
+  }
+
+  /// Create a message with the attributes from the range [@p begin, @p end)
+  template <typename Iterator>
+  MessageBuilder&& SetAttributes(Iterator begin, Iterator end) && {
+    SetAttributes(std::move(begin), std::move(end));
+    return std::move(*this);
+  }
+
+  /// Create a message with the attributes in @p v
+  MessageBuilder& SetAttributes(
+      // NOLINTNEXTLINE(performance-unnecessary-value-param)
+      std::vector<std::pair<std::string, std::string>> v) & {
```

Comment:
(no change necessary) I wonder if cases like this would work better if the argument was `absl::Span<std::pair<std::string, std::string>>`. I conceptually understand Span, but I don't have a lot of hands-on experience using it, so I'm not sure whether it's intended for cases like this not.

Reply:
Filed #4548 .